File: TailscreenProtocolPackage/Sources/TailscreenProtocol/FECGroupBuffer.swift
import Foundation

/// Receiver-side FEC state: a bounded ring of recently received video packets
/// plus briefly-buffered parity datagrams, solved via `FECCodec.recover` the
/// moment a group has exactly one member missing. Sits in front of the
/// depacketizer on the client's receive task (owned like `nackScheduler`).
///
/// Pure and deterministic — no I/O, no wall clock (the caller injects
/// `nowNs`) — so `FECGroupBufferTests` pins every path on CI. Plain
/// `Sendable` value type; the receive task is the only mutator.
public struct FECGroupBuffer: Sendable {
    /// A parity datagram whose group wasn't solvable on arrival (≥ 2 members
    /// missing, or members not yet seen — parity can outrun a reordered
    /// member). Held for one reorder tolerance, then discarded: multi-loss
    /// groups belong to NACK.
    private struct PendingParity {
        let baseSeq: UInt16
        let count: Int
        let body: Data
        let firstSeenNs: UInt64
    }

    /// Cap on retained media bytes. ~4 groups of keyframe-sized packets
    /// (~11 KB per N = 10 group) fit comfortably; keyframe bursts too.
    public let maxHeldBytes: Int
    /// Hard cap on retained media packets (second bound alongside bytes).
    public let maxHeldPackets: Int
    /// How long an unsolvable parity lingers waiting for reordered members.
    public let parityLingerNs: UInt64
    /// Cap on buffered parities (a keyframe burst holds a handful at most).
    public let maxPendingParities: Int
    /// Cap on the recovered-seq guard set (at-most-once recovery per seq).
    public let maxRecoveredGuard: Int

    /// Recently received media packets keyed by sequence number.
    private var held: [UInt16: Data] = [:]
    /// Insertion order of `held` keys, for oldest-first eviction.
    private var heldOrder: [UInt16] = []
    private var heldBytes = 0
    private var pending: [PendingParity] = []
    /// Sequence numbers already recovered once — a late original after
    /// recovery must not be double-fed by us (the original itself still
    /// flows to the depacketizer, whose reorder buffer drops it as a
    /// behind-us duplicate, same as any network dup).
    private var recovered: [UInt16] = []
    private var recoveredSet: Set<UInt16> = []

    public init(
        maxHeldBytes: Int = 256 * 1024,
        maxHeldPackets: Int = 512,
        parityLingerNs: UInt64 = 25_000_000,
        maxPendingParities: Int = 16,
        maxRecoveredGuard: Int = 256
    ) {
        self.maxHeldBytes = maxHeldBytes
        self.maxHeldPackets = maxHeldPackets
        self.parityLingerNs = parityLingerNs
        self.maxPendingParities = maxPendingParities
        self.maxRecoveredGuard = maxRecoveredGuard
    }

    /// One recovered packet, ready for the shared video-ingest path.
    public struct Recovery: Equatable {
        let seq: UInt16
        let packet: Data
    }

    /// Retain one received video packet and re-check any buffered parity
    /// whose group it may have just made solvable (parity-before-member
    /// reordering). Returns a recovery if one solved.
    public mutating func noteMedia(seq: UInt16, packet: Data, nowNs: UInt64) -> Recovery? {
        if held[seq] == nil {
            held[seq] = packet
            heldOrder.append(seq)
            heldBytes += packet.count
            evictHeldIfNeeded()
        }
        return solvePending(nowNs: nowNs)
    }

    /// Ingest one parity datagram. Solves immediately when exactly one group
    /// member is missing; drops it when none are; buffers it briefly (one
    /// reorder tolerance) when ≥ 2 are — NACK owns multi-loss groups.
    public mutating func noteParity(baseSeq: UInt16, count: Int, body: Data, nowNs: UInt64) -> Recovery? {
        purgeAgedParities(nowNs: nowNs)
        let parity = PendingParity(baseSeq: baseSeq, count: count, body: body, firstSeenNs: nowNs)
        if let recovery = trySolve(parity) {
            return recovery
        }
        if missingSeqs(baseSeq: baseSeq, count: count).isEmpty {
            return nil  // fully received — parity has nothing to add
        }
        pending.append(parity)
        if pending.count > maxPendingParities {
            pending.removeFirst(pending.count - maxPendingParities)
        }
        return nil
    }

    /// Reset for a fresh session.
    public mutating func reset() {
        held.removeAll(keepingCapacity: true)
        heldOrder.removeAll(keepingCapacity: true)
        heldBytes = 0
        pending.removeAll(keepingCapacity: true)
        recovered.removeAll(keepingCapacity: true)
        recoveredSet.removeAll(keepingCapacity: true)
    }

    // MARK: - Internals

    private func missingSeqs(baseSeq: UInt16, count: Int) -> [UInt16] {
        var missing: [UInt16] = []
        var seq = baseSeq
        for _ in 0..<count {
            if held[seq] == nil { missing.append(seq) }
            seq &+= 1
        }
        return missing
    }

    /// Attempt to solve one parity's group. Returns a recovery when exactly
    /// one member is missing, that member wasn't already recovered, and the
    /// XOR solve is internally consistent.
    private mutating func trySolve(_ parity: PendingParity) -> Recovery? {
        let missing = missingSeqs(baseSeq: parity.baseSeq, count: parity.count)
        guard missing.count == 1, let missingSeq = missing.first else { return nil }
        guard !recoveredSet.contains(missingSeq) else { return nil }

        var members: [Data] = []
        members.reserveCapacity(parity.count - 1)
        var ssrc: UInt32 = 0
        var seq = parity.baseSeq
        for _ in 0..<parity.count {
            if let member = held[seq] {
                members.append(member)
                if ssrc == 0, let (header, _) = RTPHeader.decode(from: member) {
                    ssrc = header.ssrc
                }
            }
            seq &+= 1
        }
        guard
            let packet = FECCodec.recover(
                missingSeq: missingSeq, ssrc: ssrc, members: members, body: parity.body)
        else { return nil }

        markRecovered(missingSeq)
        // The recovered packet joins the held set like a received one, so a
        // hypothetical overlapping parity sees a complete group (and drops).
        if held[missingSeq] == nil {
            held[missingSeq] = packet
            heldOrder.append(missingSeq)
            heldBytes += packet.count
            evictHeldIfNeeded()
        }
        return Recovery(seq: missingSeq, packet: packet)
    }

    /// Re-run buffered parities after a media arrival: solve the first group
    /// that just became one-missing, drop every fully-received parity, and
    /// keep the rest — a single scan over a rebuilt array, so removing a
    /// satisfied parity never skips or delays the others. At most one
    /// recovery per call (a media packet belongs to exactly one group per
    /// batch, so one arrival can complete at most one group).
    private mutating func solvePending(nowNs: UInt64) -> Recovery? {
        purgeAgedParities(nowNs: nowNs)
        guard !pending.isEmpty else { return nil }
        var recovery: Recovery?
        var remaining: [PendingParity] = []
        remaining.reserveCapacity(pending.count)
        for parity in pending {
            if recovery == nil, let solved = trySolve(parity) {
                recovery = solved  // consumed — not retained
                continue
            }
            if missingSeqs(baseSeq: parity.baseSeq, count: parity.count).isEmpty {
                continue  // fully received — parity has nothing left to add
            }
            remaining.append(parity)
        }
        pending = remaining
        return recovery
    }

    private mutating func purgeAgedParities(nowNs: UInt64) {
        pending.removeAll { nowNs &- $0.firstSeenNs > parityLingerNs }
    }

    private mutating func markRecovered(_ seq: UInt16) {
        recovered.append(seq)
        recoveredSet.insert(seq)
        if recovered.count > maxRecoveredGuard {
            let evicted = recovered.removeFirst()
            recoveredSet.remove(evicted)
        }
    }

    private mutating func evictHeldIfNeeded() {
        while heldBytes > maxHeldBytes || heldOrder.count > maxHeldPackets {
            guard !heldOrder.isEmpty else { return }
            let seq = heldOrder.removeFirst()
            if let packet = held.removeValue(forKey: seq) {
                heldBytes -= packet.count
            }
        }
    }
}

File: sdk/go/tailscreen/fecgroupbuffer.go
package tailscreen

// FECRecovery is one recovered packet, ready to be fed through the same
// ingest path as a received one (TS-FEC-011).
type FECRecovery struct {
	Seq    uint16
	Packet []byte
}

// FECGroupBufferConfig carries the buffer's bounds; zero values select the
// defaults.
type FECGroupBufferConfig struct {
	// MaxHeldBytes caps retained media bytes — about four keyframe-sized
	// groups.
	MaxHeldBytes int
	// MaxHeldPackets is the second bound alongside bytes.
	MaxHeldPackets int
	// ParityLingerNs is how long an unsolvable parity waits for reordered
	// members before multi-loss groups are left to NACK (TS-FEC-008).
	ParityLingerNs uint64
	// MaxPendingParities caps buffered parities.
	MaxPendingParities int
	// MaxRecoveredGuard caps the at-most-once-recovery guard set.
	MaxRecoveredGuard int
}

func (c FECGroupBufferConfig) withDefaults() FECGroupBufferConfig {
	if c.MaxHeldBytes == 0 {
		c.MaxHeldBytes = 256 * 1024
	}
	if c.MaxHeldPackets == 0 {
		c.MaxHeldPackets = 512
	}
	if c.ParityLingerNs == 0 {
		c.ParityLingerNs = 25_000_000
	}
	if c.MaxPendingParities == 0 {
		c.MaxPendingParities = 16
	}
	if c.MaxRecoveredGuard == 0 {
		c.MaxRecoveredGuard = 256
	}
	return c
}

type pendingParity struct {
	baseSeq     uint16
	count       int
	body        []byte
	firstSeenNs uint64
}

// FECGroupBuffer is the receiver-side FEC state: a bounded ring of recently
// received video packets plus briefly buffered parity datagrams, solved via
// Recover the moment a group has exactly one member missing (TS-FEC-008).
// It sits in front of the depacketizer.
//
// Pure and deterministic — the caller injects nowNs — and not safe for
// concurrent use; the receive path is the only mutator.
type FECGroupBuffer struct {
	cfg FECGroupBufferConfig

	held      map[uint16][]byte
	heldOrder []uint16
	heldBytes int
	pending   []pendingParity
	// recovered guards at-most-once recovery per seq: a late original after a
	// recovery must not be double-fed by us (it still flows to the
	// depacketizer, whose reorder buffer drops it as an ordinary duplicate).
	recovered    []uint16
	recoveredSet map[uint16]bool
}

// NewFECGroupBuffer constructs a buffer; zero-value config fields select the
// defaults.
func NewFECGroupBuffer(cfg FECGroupBufferConfig) *FECGroupBuffer {
	return &FECGroupBuffer{
		cfg:          cfg.withDefaults(),
		held:         make(map[uint16][]byte),
		recoveredSet: make(map[uint16]bool),
	}
}

// NoteMedia retains one received video packet and re-checks any buffered
// parity whose group it may have just made solvable — parity can outrun a
// reordered member. Returns a recovery if one solved.
func (b *FECGroupBuffer) NoteMedia(seq uint16, packet []byte, nowNs uint64) *FECRecovery {
	if _, ok := b.held[seq]; !ok {
		b.held[seq] = packet
		b.heldOrder = append(b.heldOrder, seq)
		b.heldBytes += len(packet)
		b.evictHeldIfNeeded()
	}
	return b.solvePending(nowNs)
}

// NoteParity ingests one parity datagram: solves immediately when exactly one
// group member is missing, drops it when none are, and buffers it for one
// reorder tolerance when two or more are — multi-loss groups belong to NACK.
func (b *FECGroupBuffer) NoteParity(baseSeq uint16, count int, body []byte, nowNs uint64) *FECRecovery {
	b.purgeAgedParities(nowNs)
	parity := pendingParity{baseSeq: baseSeq, count: count, body: body, firstSeenNs: nowNs}
	if recovery := b.trySolve(parity); recovery != nil {
		return recovery
	}
	if len(b.missingSeqs(baseSeq, count)) == 0 {
		return nil // fully received — the parity has nothing to add
	}
	b.pending = append(b.pending, parity)
	if len(b.pending) > b.cfg.MaxPendingParities {
		b.pending = b.pending[len(b.pending)-b.cfg.MaxPendingParities:]
	}
	return nil
}

// Reset clears all state for a fresh session.
func (b *FECGroupBuffer) Reset() {
	b.held = make(map[uint16][]byte)
	b.heldOrder = nil
	b.heldBytes = 0
	b.pending = nil
	b.recovered = nil
	b.recoveredSet = make(map[uint16]bool)
}

func (b *FECGroupBuffer) missingSeqs(baseSeq uint16, count int) []uint16 {
	var missing []uint16
	seq := baseSeq
	for i := 0; i < count; i++ {
		if _, ok := b.held[seq]; !ok {
			missing = append(missing, seq)
		}
		seq++
	}
	return missing
}

// trySolve attempts one parity's group: a recovery comes back when exactly
// one member is missing, that member was not already recovered, and the XOR
// solve is internally consistent (TS-FEC-009, TS-FEC-010).
func (b *FECGroupBuffer) trySolve(parity pendingParity) *FECRecovery {
	missing := b.missingSeqs(parity.baseSeq, parity.count)
	if len(missing) != 1 {
		return nil
	}
	missingSeq := missing[0]
	if b.recoveredSet[missingSeq] {
		return nil
	}

	members := make([][]byte, 0, parity.count-1)
	var ssrc uint32
	seq := parity.baseSeq
	for i := 0; i < parity.count; i++ {
		if member, ok := b.held[seq]; ok {
			members = append(members, member)
			if ssrc == 0 {
				if header, _, ok := DecodeRTPHeader(member); ok {
					ssrc = header.SSRC
				}
			}
		}
		seq++
	}

	packet := Recover(missingSeq, ssrc, members, parity.body)
	if packet == nil {
		return nil
	}

	b.markRecovered(missingSeq)
	// The recovered packet joins the held set like a received one, so an
	// overlapping parity sees a complete group and drops.
	if _, ok := b.held[missingSeq]; !ok {
		b.held[missingSeq] = packet
		b.heldOrder = append(b.heldOrder, missingSeq)
		b.heldBytes += len(packet)
		b.evictHeldIfNeeded()
	}
	return &FECRecovery{Seq: missingSeq, Packet: packet}
}

// solvePending re-runs buffered parities after a media arrival: solve the
// first group that just became one-missing, drop every fully received
// parity, and keep the rest — one scan over a rebuilt slice, so removing a
// satisfied parity never skips or delays the others. At most one recovery
// per call: a media packet belongs to exactly one group per batch, so one
// arrival can complete at most one group.
func (b *FECGroupBuffer) solvePending(nowNs uint64) *FECRecovery {
	b.purgeAgedParities(nowNs)
	if len(b.pending) == 0 {
		return nil
	}
	var recovery *FECRecovery
	remaining := b.pending[:0]
	for _, parity := range b.pending {
		if recovery == nil {
			if solved := b.trySolve(parity); solved != nil {
				recovery = solved // consumed — not retained
				continue
			}
		}
		if len(b.missingSeqs(parity.baseSeq, parity.count)) == 0 {
			continue // fully received — nothing left to add
		}
		remaining = append(remaining, parity)
	}
	b.pending = remaining
	return recovery
}

func (b *FECGroupBuffer) purgeAgedParities(nowNs uint64) {
	kept := b.pending[:0]
	for _, parity := range b.pending {
		if nowNs-parity.firstSeenNs <= b.cfg.ParityLingerNs {
			kept = append(kept, parity)
		}
	}
	b.pending = kept
}

func (b *FECGroupBuffer) markRecovered(seq uint16) {
	b.recovered = append(b.recovered, seq)
	b.recoveredSet[seq] = true
	if len(b.recovered) > b.cfg.MaxRecoveredGuard {
		evicted := b.recovered[0]
		b.recovered = b.recovered[1:]
		delete(b.recoveredSet, evicted)
	}
}

func (b *FECGroupBuffer) evictHeldIfNeeded() {
	for (b.heldBytes > b.cfg.MaxHeldBytes || len(b.heldOrder) > b.cfg.MaxHeldPackets) && len(b.heldOrder) > 0 {
		seq := b.heldOrder[0]
		b.heldOrder = b.heldOrder[1:]
		if packet, ok := b.held[seq]; ok {
			delete(b.held, seq)
			b.heldBytes -= len(packet)
		}
	}
}
